src/circuit_builder.rs

```diff
@@ -38,44 +273,162 @@ impl PublicInput {
 
 /// Circuit representation for a gadget with all of the tools that it
 /// should implement.
-pub trait Circuit<'a>
-where
-    Self: Sized + Default,
-{
+pub trait Circuit: Sized {
     /// Gadget implementation used to fill the composer.
-    fn gadget(&mut self, composer: &mut StandardComposer) -> Result<Vec<PublicInput>>;
-    /// Compiles the circuit by using a function that returns a `Result`
-    /// with the `ProverKey`, `VerifierKey` and the circuit size.
-    fn compile(&mut self, pub_params: &PublicParameters)
-        -> Result<(ProverKey, VerifierKey, usize)>;
+    fn gadget(&mut self, composer: &mut StandardComposer) -> Result<Vec<PublicInput>, Error>;
+
+    /// Compiles the circuit.
+    ///
+    /// Set the commit, opening, prover and verifier keys.
+    ///
+    /// Also, set the circuit size.
+    fn compile(&mut self, pub_params: &PublicParameters) -> Result<(), Error> {
+        let (ck, ok) = pub_params.trim(Self::max_degree())?;
+
+        self.set_commit_key(ck.clone());
+        self.set_opening_key(ok);
+
+        let mut prover = Prover::new(Self::compile_label());
+        let mut verifier = Verifier::new(Self::compile_label());
+
+        self.gadget(prover.mut_cs())?;
```

Comment:
Pi constructor should be set here for self.

Reply:
The compilation of the circuit is required only to set the reference string (so we can later on derive the opening, commit, prover and verifier keys)

If we set the PI constructor here, it implies we will need to serialize / deserialize additional data when compiling circuits.